### DISCUSSION TITLE
An AI Agent that write queries to the data source based on a natural language question

### DISCUSSION THREAD

**Discussion by samber** [grafana/grafana#71618] [Plugins]
What about a little companion for writing PromQL, SQL, Elasticsearch... queries faster, based on the actual data source schema?

**Example:**

The user type the following question: "http request rate per API version for api.acme.org endpoint".

Grafana would send it to a LLM (GPT4?) with all Prometheus metric names+descriptions. As soon as the LLM replies with a few metrics matching the request, a new LLM query could be sent, containing the labels and some value samples. Then, the returned PromQL query is inserted into the panel.

Selecting a data source dynamically or even the panel type would be awesome.

With an advanced prompt, I'm pretty sure it would be possible to prefill the query with board variables.

Metabase built a similar feature recently: Metabot.

**Comment by tedraymond** [comment]
I've tested the recend chatgpt with LLM plugin. Sadly it only allows for OpenAI or Azure.

Curious, has anyone attempted connecting to a local onprem Private GPT or LLM?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82586. Feel free to continue the discussion around this there. Thank you!
